### PR TITLE
[Misc] Support FP8 MoE for compressed-tensors

### PR DESCRIPTION
**Pull request by mgoin**
Adds support for loading compressed-tensors format MoE models that have been compressed using FP8 per-tensor weights and activations, leveraging the existing kernel

Models tested (GSM8k eval is on the model card):
* [x] Mixtral https://huggingface.co/mgoin/Mixtral-8x7B-Instruct-v0.1-FP8
* [x] DeepSeek https://huggingface.co/mgoin/DeepSeek-Coder-V2-Lite-Instruct-FP8
* [x] OLMoE https://huggingface.co/nm-testing/OLMoE-1B-7B-0924-Instruct-FP8
* [x] Phi3MoE https://huggingface.co/nm-testing/Phi-3.5-MoE-instruct-FP8 (this required fixing the model def)

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 1]
```diff
@@ -353,12 +353,13 @@ def weight_loader(self, param: torch.nn.Parameter,
...
-                    "input_scales of w1 and w3 of a layer "
-                    f"must be equal. But got {param.data[expert_id]} "
-                    f"vs. {loaded_weight}")
+            # print(param.data.device, loaded_weight.device)
+            # if param.data[expert_id] != 1 and (param.data[expert_id] -
+            #                                    loaded_weight).abs() > 1e-5:
+            #     raise ValueError(
+            #         "input_scales of w1 and w3 of a layer "
+            #         f"must be equal. But got {param.data[expert_id]} "
+            #         f"vs. {loaded_weight}")
```
[the remark this answers is not in the document]
Comment (mgoin) [reply vllm/model_executor/layers/fused_moe/layer.py]: It is still needed at the moment for all fp8 cases since we only support one scale for the fused weights

**Comment by mgoin** [comment]
It turns out Phi3 wasn't working because quant_config wasn't being passed to qkv_proj and o_proj, very straightforward

It works for Phi3MoE
```
vllm (pretrained=nm-testing/Phi-3.5-MoE-instruct-FP8,trust_remote_code=True,max_model_len=4096,gpu_memory_utilization=0.8), gen_kwargs: (None), limit: None, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8681|±  |0.0093|
|     |       |strict-match    |     5|exact_match|↑  |0.8431|±  |0.0100|
```

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 2]
```diff
@@ -353,6 +355,8 @@ def weight_loader(self, param: torch.nn.Parameter,
 
         # Case input scale: input_scale loading is only supported for fp8
         if "input_scale" in weight_name:
+            loaded_weight = loaded_weight.to(param.data.device)
```
Comment (dsikka) [inline vllm/model_executor/layers/fused_moe/layer.py:358]: Can you leave a note here that this is needed for compressed-tensors only since we also support input scales for autoFP8

File: tests/weight_loading/models.txt [hunk 3]
```diff
@@ -20,6 +20,7 @@ compressed-tensors, nm-testing/Meta-Llama-3-8B-FP8-compressed-tensors-test, main
 compressed-tensors, nm-testing/Phi-3-mini-128k-instruct-FP8, main
 compressed-tensors, neuralmagic/Phi-3-medium-128k-instruct-quantized.w4a16, main
 compressed-tensors, nm-testing/TinyLlama-1.1B-Chat-v1.0-actorder-group, main
+compressed-tensors, mgoin/DeepSeek-Coder-V2-Lite-Instruct-FP8, main
```
Comment (dsikka) [inline tests/weight_loading/models.txt:23]: could move to the large models list as well

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 4]
```diff
@@ -73,7 +73,17 @@ def get_quant_method(
         if isinstance(layer, Attention):
             return CompressedTensorsKVCacheMethod(self)
         if isinstance(layer, FusedMoE):
-            return CompressedTensorsMoEMethod(self)
+            # TODO: @dsikka: refactor this to use schemes as other kernels
+            # are supported + check if the layer is being ignored.
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:77]: Since we're now expanding the supported schemes, we should have this just return `CompressedTensorsMoEMethod` and have the return the correct MoE scheme within the class since it looks like we're now checking the conditions twice?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 5]
```diff
@@ -5,23 +5,219 @@
...
+        self.weight_quant = self.quant_config.target_scheme_map["Linear"].get(
+            "weights")
+        self.input_quant = self.quant_config.target_scheme_map["Linear"].get(
+            "input_activations")
+
+        assert self.weight_quant.symmetric and self.input_quant.symmetric, (
+            "Only symmetric quantization is supported for MoE")
+        assert self.weight_quant.strategy.value == "tensor"
+        assert self.input_quant.strategy.value == "tensor"
+        assert not self.weight_quant.dynamic
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:53]: Can we raise an error here if the conditions are met, clearly stating what is currently supported

**Review by dsikka — APPROVED** [review]
LGTM. Triggered the weight loading large model test which will run all the MoE models.
